src/FactSystem/FactMetaData.cc

```diff
@@ -1012,67 +1012,34 @@ const FactMetaData::AppSettingsTranslation_s* FactMetaData::_findAppSettingsHori
             continue;
         }
 
-        uint settingsUnits = qgcApp()->toolbox()->settingsManager()->unitsSettings()->horizontalDistanceUnits()->rawValue().toUInt();
-
-        if (pAppSettingsTranslation->unitType == UnitHorizontalDistance
-                && pAppSettingsTranslation->unitOption == settingsUnits) {
-            return pAppSettingsTranslation;
-        }
-    }
-    return nullptr;
-}
-
-const FactMetaData::AppSettingsTranslation_s* FactMetaData::_findAppSettingsVerticalDistanceUnitsTranslation(const QString& rawUnits)
-{
-    for (size_t i=0; i<sizeof(_rgAppSettingsTranslations)/sizeof(_rgAppSettingsTranslations[0]); i++) {
-        const AppSettingsTranslation_s* pAppSettingsTranslation = &_rgAppSettingsTranslations[i];
-
-        if (rawUnits.toLower() != pAppSettingsTranslation->rawUnits.toLower()) {
-            continue;
-        }
-
-        uint settingsUnits = qgcApp()->toolbox()->settingsManager()->unitsSettings()->verticalDistanceUnits()->rawValue().toUInt();
-
-        if (pAppSettingsTranslation->unitType == UnitVerticalDistance
-                && pAppSettingsTranslation->unitOption == settingsUnits) {
-            return pAppSettingsTranslation;
-        }
-    }
-    return nullptr;
-}
-
-const FactMetaData::AppSettingsTranslation_s* FactMetaData::_findAppSettingsWeightUnitsTranslation(const QString& rawUnits)
-{
-    for (size_t i=0; i<sizeof(_rgAppSettingsTranslations)/sizeof(_rgAppSettingsTranslations[0]); i++) {
-        const AppSettingsTranslation_s* pAppSettingsTranslation = &_rgAppSettingsTranslations[i];
-
-        if (rawUnits.toLower() != pAppSettingsTranslation->rawUnits.toLower()) {
-            continue;
+        uint unitOption = 0;
+        auto unitsSettings = qgcApp()->toolbox()->settingsManager()->unitsSettings();
+        switch (type) {
+        case UnitHorizontalDistance:
+            unitOption = unitsSettings->horizontalDistanceUnits()->rawValue().toUInt();
+            break;
+        case UnitVerticalDistance:
+            unitOption = unitsSettings->verticalDistanceUnits()->rawValue().toUInt();
+            break;
+        case UnitArea:
+            unitOption = unitsSettings->areaUnits()->rawValue().toUInt();
+            break;
+        case UnitSpeed:
+            unitOption = unitsSettings->speedUnits()->rawValue().toUInt();
+            break;
+        case UnitTemperature:
+            unitOption = unitsSettings->temperatureUnits()->rawValue().toUInt();
+            break;
+        case UnitWeight:
+            unitOption = unitsSettings->weightUnits()->rawValue().toUInt();
+            break;
+        default:
+            assert(false);
```

Comment:
Oops sorry, just saw this now. You can't used asserts in QGC code. That causes the app to crash, which is never a good idea for a GCS. For switch statements QGC compiles will fail build if not all cases are covered and there is no default. So there is no need for a default here. If a new unit gets added later the build will fail until this switch is updated.

Also just FYI: The way to handle switch statements for which a default case is needed since not all the cases over covered. And for which that default indicates an error, the general pattern is to use qWarning with text spitting out "Internal Error: ..." to indicate a coding error.

Reply:
We used assert not to indicate all cases are covered or not but to check if no other value will be passed which is an error. Additional program logic check. Above `_setAppSettingsTranslators()` method contains default for some reason which is not correct. But ok, changed to your coding standard: https://github.com/mavlink/qgroundcontrol/pull/8820/commits/8a33a574431d0249f7ab3236a035d00aceafa97f

Reply:
"We used assert not to indicate all cases are covered or not but to check if no other value will be passed which is an error. "

The variable is typed as UnitsType which means it will be a compiler error to get anything else. I guess you could cast a crap value to a UnitsType which a default will catch. But to me, it's better to have the compiler complain when a new type is added to let you know the switch must be updated than handle that case.

Reply:
while enums widely used in QML we won't get compiler errors in many cases ;)

Reply:
The way to catch something like that is to validate the value at the point of entry with separate code. That way you get both validation and compiler errors with new enums added. As it currently stands QGC does not generally validate enums which come in from the Qml side. It unfortunately expects Qml code to be correct which isn't the best thing.